service/matching/taskQueueManager.go

```diff
@@ -479,14 +482,31 @@ func (c *taskQueueManagerImpl) GetUserData(ctx context.Context) (*persistencespb
 	return data, err
 }
 
-func (c *taskQueueManagerImpl) UpdateUserData(ctx context.Context, updateFn func(*persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, error)) error {
+//nolint:revive // control coupling
+func (c *taskQueueManagerImpl) UpdateUserData(ctx context.Context, replicate bool, updateFn func(*persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, error)) error {
 	newData, err := c.db.UpdateUserData(ctx, updateFn)
 	c.signalIfFatal(err)
 	if err != nil {
 		return err
 	}
+	if replicate && c.namespaceReplicationQueue != nil {
+		err = c.namespaceReplicationQueue.Publish(ctx, &repication.ReplicationTask{
+			TaskType: enumsspb.REPLICATION_TASK_TYPE_TASK_QUEUE_USER_DATA,
+			Attributes: &repication.ReplicationTask_TaskQueueUserDataAttributes{
+				TaskQueueUserDataAttributes: &repication.TaskQueueUserDataAttributes{
+					NamespaceId:   c.taskQueueID.namespaceID.String(),
+					TaskQueueName: c.taskQueueID.BaseNameString(),
+					UserData:      newData.GetData(),
+				},
+			},
+		})
+		if err != nil {
+			c.logger.Error("Failed to publish a replication task after updating task queue user data", tag.Error(err))
+			return serviceerror.NewFailedPrecondition("Storing task queue user data succeeded but publishing to the namespace replication queue failed, please try again")
```

Comment:
i would use unavailable error here,
failed precondition should be used if API is explicitly doing CAS, e.g. only update the user record if the record is not touched